### PR TITLE
Add clang/docs/FunctionEffectAnalysis.rst.

### PR DESCRIPTION
**Pull request by dougsonos**
Follow-on from #99656, which introduces 2nd pass caller/callee analysis for function effects.

Wrote a new documentation page, derived directly from the RFC posted to LLVM Discourse earlier this year.

### CODE REVIEW

File: clang/docs/FunctionEffectAnalysis.rst
Comment (cjappl) [inline clang/docs/FunctionEffectAnalysis.rst:8]: It is not only C++, but also works in C I think?

Comment (cjappl) [reply clang/docs/FunctionEffectAnalysis.rst:8]: May be worth documenting how to use in C, I know the syntax is a little different with the attrs

File: clang/docs/FunctionEffectAnalysis.rst
Comment (cjappl) [inline clang/docs/FunctionEffectAnalysis.rst:93]: nit/style (take it or leave it) - some commas could be removed

File: clang/docs/FunctionEffectAnalysis.rst
Comment (cjappl) [inline clang/docs/FunctionEffectAnalysis.rst:272]: I think more appropriately "the compiler may generate a lock" -- I'm not sure it is always guaranteed to generate a lock, but will in some circumstances. Someone with more experience than me can confirm or deny

**Review by cjappl — COMMENTED** [review]
(partial review, will do another pass later today -- overall looks great so far)

File: clang/docs/FunctionEffectAnalysis.rst
Comment (cjappl) [inline clang/docs/FunctionEffectAnalysis.rst:59]: It may be good to mention these also control RealtimeSanitizer and do a link to our documentation page.

After this goes in I can submit a PR to cross link from that page here, as I think it's beneficial to the user to see the connection seeing as the attrs are shared

File: clang/docs/FunctionEffectAnalysis.rst
Comment (cjappl) [inline clang/docs/FunctionEffectAnalysis.rst:103]: nit (take it or leave it): This may possibly be clearer to show in a code block with the warnings emitted? That matches the style in the rest of the doc and also gives good search strings for the user wondering what's going wrong

File: clang/docs/FunctionEffectAnalysis.rst
Comment (cjappl) [inline clang/docs/FunctionEffectAnalysis.rst:241]: One question that doesn't necessarily pertain to this PR -- what about user overloads of new and delete? If I overload new and delete to go to some nonblocking memory pool, is there a way I can mark it as safe?

File: clang/docs/FunctionEffectAnalysis.rst
Comment (cjappl) [inline clang/docs/FunctionEffectAnalysis.rst:333]: "It is possible for a non-blocking function to schedule..."

lock-free possibly left over from old draft

File: clang/docs/FunctionEffectAnalysis.rst
Comment (cjappl) [inline clang/docs/FunctionEffectAnalysis.rst:351]: nit: May consider moving this further up the document, I know as a user when I crack open these docs my first questions are: how do I use it (basically) and how do I disable it.

The closer this information is to the top of the doc, the better

File: clang/docs/FunctionEffectAnalysis.rst
Comment (cjappl) [inline clang/docs/FunctionEffectAnalysis.rst:394]: nice, this is helpful

File: clang/docs/FunctionEffectAnalysis.rst
Comment (cjappl) [inline clang/docs/FunctionEffectAnalysis.rst:408]: Not sure if it is worth mentioning, but most std::function implementations (all?) allocate under some circumstances opaque to the user, so even if the attribute was not lost it is still unsafe.

**Review by cjappl — APPROVED** [review]
Overall, very thorough. I would wait for more eyes but this LGTM

File: clang/docs/FunctionEffectAnalysis.rst
Comment (Sirraide) [reply clang/docs/FunctionEffectAnalysis.rst:408]: Yeah, afaik, `std::function` is a bit like `std::string` in that regard: if the function pointer (+ its context if it’s a lambda) is small enough (whatever that means for any one implementation), it’s stored inline; otherwise, it has to allocate.

Comment (Sirraide) [reply clang/docs/FunctionEffectAnalysis.rst:408]: So yeah, not sure option 2 below is even *possible* in the general case—it certainly isn’t for all possible lambdas.

I’d also add something like `llvm::function_ref` (iirc `std::function_ref` is also a thing since C++26 and does the exact same thing) as another option: it doesn’t own the underlying callable and thus doesn’t have to store it or allocate anything. It’s basically to `std::function` what `std::string_view` is to `std::string`. That works pretty well for function parameters, but it falls apart if you actually need to store the callable, of course.

File: clang/docs/FunctionEffectAnalysis.rst
Comment (Sirraide) [inline clang/docs/FunctionEffectAnalysis.rst:500]: I think we generally just treat most C standard library functions as builtins, not just `<cmath>`, don’t we?

File: clang/docs/FunctionEffectAnalysis.rst
Comment (dougsonos) [reply clang/docs/FunctionEffectAnalysis.rst:93]: @Sirraide, despite my being a former English major who dropped out to hack on computers and play in bands, I don't mind having my writing picked apart at all! Thanks :)

Comment (dougsonos) [reply clang/docs/FunctionEffectAnalysis.rst:93]: (Though, I am in the middle of incorporating this round of feedback + fixing whatever else looks wrong...)

File: clang/docs/FunctionEffectAnalysis.rst
Comment (dougsonos) [reply clang/docs/FunctionEffectAnalysis.rst:214]: I too was surprised by this when testing various bits of this document against the implementation yesterday.

```
/tmp/foo.cpp:9:2: error: call to 'f3' is ambiguous
    9 |         f3(nullptr);
      |         ^~
/tmp/foo.cpp:1:8: note: candidate function
    1 |   void f3(void (*)()) {}
      |        ^
/tmp/foo.cpp:2:8: note: candidate function
    2 |   void f3(void (*)() noexcept) {}
      |        ^
/tmp/foo.cpp:10:2: error: call to 'f4' is ambiguous
   10 |         f4(nullptr);
      |         ^~
/tmp/foo.cpp:4:8: note: candidate function
    4 |   void f4(void (*)()) {}
      |        ^
/tmp/foo.cpp:5:8: note: candidate function
    5 |   void f4(void (*)() [[clang::nonblocking]]) {}
      |        ^
2 errors generated.
```

We do not have a test for this specific case but I could add one.

Comment (dougsonos) [reply clang/docs/FunctionEffectAnalysis.rst:214]: There is however this test:
```c++
// --- OVERLOADS ---
#ifdef __cplusplus
struct S {
  void foo(); // expected-note {{previous declaration is here}}
  void foo() [[clang::nonblocking]]; // expected-error {{class member cannot be redeclared}}
};
#endif // __cplusplus
```

File: clang/docs/FunctionEffectAnalysis.rst
Comment (dougsonos) [reply clang/docs/FunctionEffectAnalysis.rst:248]: Ah I was confused by the `std::string` constructor parameters to the exception classes, thanks.

File: clang/docs/FunctionEffectAnalysis.rst
Comment (dougsonos) [reply clang/docs/FunctionEffectAnalysis.rst:272]: Yeah, when I looked at this a long time ago, I found that the decision in the back-end to emit a lock (or not) didn't seem trivially (ha) replicable in Sema.

File: clang/docs/FunctionEffectAnalysis.rst
Comment (dougsonos) [reply clang/docs/FunctionEffectAnalysis.rst:351]: I'm kinda keen on a natural order of exposition. But maybe a table of contents will address this?

image

File: clang/docs/FunctionEffectAnalysis.rst
Comment (Sirraide) [reply clang/docs/FunctionEffectAnalysis.rst:214]: I’d say add a test for that, yeah (including one that passes for both overloads); doesn’t have to be part of the part 2 pr; that can also be separate.

File: clang/docs/FunctionEffectAnalysis.rst
Comment (Sirraide) [reply clang/docs/FunctionEffectAnalysis.rst:351]: Hmm, that might help yeah

File: clang/docs/FunctionEffectAnalysis.rst
Comment (dougsonos) [reply clang/docs/FunctionEffectAnalysis.rst:500]: Yeah, math.h is just the first cluster of examples I ran into in adoption...

File: clang/docs/FunctionEffectAnalysis.rst
Comment (cjappl) [reply clang/docs/FunctionEffectAnalysis.rst]: I just discovered this table format when updating our docs, which is way way faster and more sustainable to write:

```
.. list-table:: Runtime Flags
   :widths: 20 10 70
   :header-rows: 1

   * - Flag name
     - Default value
     - Short description
   * - halt_on_error
     - true
     - Exit after first reported error.
   * - print_stats_on_exit
     - false
     - Exit after first reported error. If false, deduplicates error stacks so errors only appear once.
   * - color
     - auto
     - Colorize reports: (always|never|auto).
   * - fast_unwind_on_fatal
     - false
     - If available, use the fast frame-pointer-based unwinder on detected errors.

```

Saved me a lot of adjusting/readjusting the "cells" by hand.

**Comment by dougsonos** [comment]
Ping

**Comment by dougsonos** [comment]
Ping

File: clang/docs/FunctionEffectAnalysis.rst
Comment (Sirraide) [inline clang/docs/FunctionEffectAnalysis.rst:101]:
```suggestion
``nonallocating(false)`` are legal, but superfluous  when applied to a function *type*
that is not part of a declarator: ``float (int) [[nonblocking(false)]]`` and 
``float (int)`` are identical types.
```
Just because that might not be 100% clear otherwise.

File: clang/docs/FunctionEffectAnalysis.rst
Comment (Sirraide) [inline clang/docs/FunctionEffectAnalysis.rst:128]:
```suggestion
or strengthen a performance constraint is unsafe and results in a warning. The rules for this
are comparable to that for ``noexcept`` in C++17 and later.
```
That might help people who are familiar w/ how `noexcept` works.

File: clang/docs/FunctionEffectAnalysis.rst
Comment (Sirraide) [inline clang/docs/FunctionEffectAnalysis.rst:165]:
```suggestion
In C++, when a virtual method has a performance constraint, overriding methods in
```
‘overriding methods in subclasses’ already makes it clear what’s happening here.

File: clang/docs/FunctionEffectAnalysis.rst
Comment (Sirraide) [inline clang/docs/FunctionEffectAnalysis.rst:199]:
```suggestion
Unlike ``noexcept``, a redeclaration of ``f2`` with an added or stronger performance constraint is
legal and propagates the attribute to the previous declaration:
```

File: clang/docs/FunctionEffectAnalysis.rst
Comment (Sirraide) [inline clang/docs/FunctionEffectAnalysis.rst:206]:
```suggestion
This greatly eases adoption by making it possible to annotate functions in external libraries
```

File: clang/docs/FunctionEffectAnalysis.rst
Comment (Sirraide) [inline clang/docs/FunctionEffectAnalysis.rst:210]:
```suggestion
A redeclaration with a removed or weaker performance constraint produces a warning, paralleling
the behavior of ``noexcept``:
```
‘in order to’ is technically wrong: we don’t warn ‘just because `noexcept` does’, but because it makes sense to warn on it ;Þ

File: clang/docs/FunctionEffectAnalysis.rst
Comment (Sirraide) [inline clang/docs/FunctionEffectAnalysis.rst:241]: I think we can just remove this section entirely? Didn’t we already pretty much state exactly this above where we mentioned the `-Wperf-constraint-implies-noexcept` warning?

File: clang/docs/FunctionEffectAnalysis.rst
Comment (Sirraide) [inline clang/docs/FunctionEffectAnalysis.rst:278]:
```suggestion
     form of the required attribute, and can be verified to have the same attribute or stronger,
```

File: clang/docs/FunctionEffectAnalysis.rst
Comment (Sirraide) [inline clang/docs/FunctionEffectAnalysis.rst:327]:
```suggestion
a function that is called from a performance-constrained function may be analyzed to
```

File: clang/docs/FunctionEffectAnalysis.rst
Comment (Sirraide) [inline clang/docs/FunctionEffectAnalysis.rst:393]:
```suggestion
- constructs which do block, but which in practice are used in ways to avoid unbounded blocking,
  e.g. a thread pool with semaphores to coordinate multiple realtime threads;
- using libraries which are safe but not yet annotated;
```

File: clang/docs/FunctionEffectAnalysis.rst
Comment (Sirraide) [inline clang/docs/FunctionEffectAnalysis.rst:423]:
```suggestion
code, because a typical implementation may allocate heap memory in the constructor.
```
Iirc most implementations of `std::function` make use of small-buffer optimisation like `std::string`, so function pointers, non-capturing lambdas, etc. can be stored inline.

File: clang/docs/FunctionEffectAnalysis.rst
Comment (Sirraide) [inline clang/docs/FunctionEffectAnalysis.rst:541]:
```suggestion
It also helps that many of the functions in the standard C libraries (notably ``<math.h>``)
are treated as built-in functions by Clang, which the diagnosis understands to be safe.
```

**Review by Sirraide — COMMENTED** [review]
Just gave the entire thing another read, and I have a few nits and comments, but other than that, I’m personally fine w/ it.

**Comment by dougsonos** [comment]
@Sirraide, thanks for the excellent feedback. I've incorporated all of your suggestions.

**Comment by cjappl** [comment]
@dougsonos let me know if you want to merge this, happy to push the button when you are OK with it

**Review by Sirraide — APPROVED** [review]

**Comment by dougsonos** [comment]
@cjappl please merge at will. Thank you!

**Comment by cjappl** [comment]
Happy to help. I'll merge this afternoon after the checks are green. 

Excited to have these docs up, had many conversations where I've wanted to refer people to them!

**Comment by Sirraide** [comment]
@dougsonos Now that this document exists, we should also update the function effects release note that the part 2 pr added to include a link to it. 👀

**Pull request merged**
